* Empirical distribution mixture models

# TODO: Add citation.

# For your reference, please see and cite CITATION.

Empirical distribution mixture (EDM) models assume that evolution occurs along a
species tree according to a mixture of amino acid substitution models. The
transition rate matrices of the different components share one set of
exchangeabilities. In contrast, the stationary distributions differ between each
component of the mixture model. The Python script presented here obtains
stationary distributions with corresponding weights from a set of site
distributions. A site distribution is the probability distribution of used amino
acids at a specific site in the alignment. Consequently, site distributionare
are points in 20-dimensional space with elements summing up to 1.0.

Site distributions can be calculated during Bayesian analysis, for example with
Phylobayes and the CAT model [1] . In this case, a site distribution corresponds
to the expectation of the posterior distribution of the stationary distribution
of amino acids at a specific site.

The script presented here can transform the site distributions before further
analysis. Application of linear transformations is a standard procedure in
analyses of compositional data. The two provided transformations are: (1) the
centered log ratio transformation (CLR, [2]), and (2) the log centered log ratio
transformation (LCLR, [3]). The transformations ensure that site distributions
exhibiting specific different features are moved further apart from each other,
so that they fall into different groups in the subsequent analysis.

K-means clustering with Scikit-Learn [4] is used to group the prepared site
distributions into 4, 8, 16, 32, 64, 128, and 256 clusters. The cluster centers,
which are the stationary distributions, and weights can be used as components of
EDM models.

Further, a set of stationary distributions and weights estimated from sub-sets
of the HOGENOM [5] and HSSP [6] data bases, as well as a union of both are
provided. The files can directly be used with several established phylogenetic
softwware packages such as IQ-TREE [7], Phylobayes [8], or RevBayes [9].

# TODO: Provide command lines for the three software packages.

#+NAME: Help
#+BEGIN_SRC sh :exports results :results output verbatim
./edm-cluster --help
#+END_SRC

#+RESULTS: Help
#+begin_example
usage: edm-cluster [-h] [--transformation {none,clr,lclr}] [-k K] [-p PREFIX]
                   filenames [filenames ...]

Transform and cluster site distributions. Each amino acid frequency
distribution distribution is considered as a point in twenty dimensional space.
The K-Means clustering algorithm is used to label the points from 1 to K, with
different values of K. The clusters are output in various file formats, readily
usable with IQ-TREE [1], Phylobayes [2], and RevBayes [3].

Before clustering, coordinate transformations can be applied.

Implemented transformations:
- centered log ratio (CLR) transform [4]
- log centered log ratio (LCLR) transform [5]

Basically, points close to the boundaries, that is, where some coordinates have
very low probability, will be projected to a position that is very far away
from points that have high diversity and intermediate frequency values for all
coordinates.

The site distribution files are expected to be tab separated, and of the
following form:

----------------------------------------------------------------------
A C D E F G H I K L M N P Q R S T V W Y

1	0.008	0.001	0.001	0.007	0.002	0.004	0.01	0.006	0.14	0.006	0.005	0.009	0.002	0.01	0.7	0.009	0.01	0.006	0.002	0.003
2	...
...
----------------------------------------------------------------------

Empirical distribution mixture models. Developed by Dominik Schrempf (2019).

positional arguments:
  filenames             Names of files containing site distributions

optional arguments:
  -h, --help            show this help message and exit
  --transformation {none,clr,lclr}, -t {none,clr,lclr}
                        Transformation (default: None)
  -k K                  Number of clusters (default: 64)
  -p PREFIX, --prefix PREFIX
                        Prefix of output files. (default: None)

[1] Nguyen, L., Schmidt, H. A., von Haeseler, A., & Minh, B. Q. (2015).
Iq-tree: a fast and effective stochastic algorithm for estimating
maximum-likelihood phylogenies. Molecular Biology and Evolution, 32(1),
268–274. http://dx.doi.org/10.1093/molbev/msu300

[2] Lartillot, N., Rodrigue, N., Stubbs, D., & Richer, J. (2013). Phylobayes
mpi: phylogenetic reconstruction with infinite mixtures of profiles in a
parallel environment. Systematic Biology, 62(4), 611–615.
http://dx.doi.org/10.1093/sysbio/syt022

[3] Höhna, S., Landis, M. J., Heath, T. A., Boussau, B., Lartillot, N., Moore,
B. R., Huelsenbeck, J. P., … (2016). Revbayes: bayesian phylogenetic inference
using graphical models and an interactive model-specification language.
Systematic Biology, 65(4), 726–736. http://dx.doi.org/10.1093/sysbio/syw021

[4] Aitchison, J. (1982). The statistical analysis of compositional data.
Journal of the Royal Statistical Society, Series B (Methological), 44(2),
139–177.

[5] Godichon-Baggioni, A., Maugis-Rabusseau, C., & Rau, A. (2017).
Clustering transformed compositional data using K-means, with applications in
gene expression and bicycle sharing system data. ArXiv, 1–32.
#+end_example

* Footnotes

[1] Lartillot, N., & Philippe, H., A bayesian mixture model for across-site
heterogeneities in the amino-acid replacement process, Molecular Biology and
Evolution, 21(6), 1095–1109 (2004). http://dx.doi.org/10.1093/molbev/msh112

[2] Aitchison, J., The statistical analysis of compositional data, Journal of
the Royal Statistical Society, Series B (Methological), 44(2), 139–177 (1982).

[3] Godichon-Baggioni, A., Maugis-Rabusseau, C., & Rau, A., Clustering
transformed compositional data using k-means, with applications in gene
expression and bicycle sharing system data, Journal of Applied Statistics,
46(1), 47–65 (2018). http://dx.doi.org/10.1080/02664763.2018.1454894

[4] Pedregosa, F., Varoquaux, G., Gramfort, A., Michel, V., Thirion, B., Grisel,
O., Blondel, M., …, Scikit-learn: machine learning in Python, Journal of Machine
Learning Research, 12(), 2825–2830 (2011).

[5] Dufayard, J., Duret, L., Penel, S., Gouy, M., Rechenmann, F., & Perrière,
G., Tree pattern matching in phylogenetic trees: automatic search for orthologs
or paralogs in homologous gene sequence databases, Bioinformatics, 21(11),
2596–2603 (2005). http://dx.doi.org/10.1093/bioinformatics/bti325

[6] Schneider, R., Daruvar, A. d., & Sander, C., The HSSP database of protein
structure-sequence alignments, Nucleic Acids Research, 25(1), 226–230 (1997).
http://dx.doi.org/10.1093/nar/25.1.226

[7] Nguyen, L., Schmidt, H. A., von Haeseler, A., & Minh, B. Q., Iq-tree: a fast
and effective stochastic algorithm for estimating maximum-likelihood
phylogenies, Molecular Biology and Evolution, 32(1), 268–274 (2015).
http://dx.doi.org/10.1093/molbev/msu300

[8] Lartillot, N., Rodrigue, N., Stubbs, D., & Richer, J., Phylobayes mpi:
phylogenetic reconstruction with infinite mixtures of profiles in a parallel
environment, Systematic Biology, 62(4), 611–615 (2013).
http://dx.doi.org/10.1093/sysbio/syt022

[9] Höhna, S., Landis, M. J., Heath, T. A., Boussau, B., Lartillot, N., Moore,
B. R., Huelsenbeck, J. P., …, Revbayes: bayesian phylogenetic inference using
graphical models and an interactive model-specification language, Systematic
Biology, 65(4), 726–736 (2016). http://dx.doi.org/10.1093/sysbio/syw021







