* Empirical distribution mixutre models

# TODO: Add citation.

#+NAME: Help
#+BEGIN_SRC sh :exports results :results output verbatim
./cluster-site-distributions --help
#+END_SRC

#+RESULTS: Help
#+begin_example
usage: cluster-site-distributions [-h] [--transformation {none,clr,lclr}]
                                  [-k K] [-p PREFIX]
                                  filenames [filenames ...]

Cluster site distributions. Each amino acid frequency distribution
distribution is considered as a point in twenty dimensional space. The K-Means
clustering algorithm is used to label the points from 1 to K, with different
values of K. The clusters are output in various file formats, readily usable
with IQ-TREE [1], Phylobayes [2], and RevBayes [3].

Before clustering, coordinate transformations can be applied.

Implemented transformations:
- centered log ratio (CLR) transform [4]
- log centered log ratio (LCLR) transform [5]

Basically, points close to the boundaries, that is, where some coordinates have
very low probability, will be projected to a position that is very far away
from points that have high diversity and intermediate frequency values for all
coordinates.

The site distribution files are expected to be tab separated, and of the
following form:

----------------------------------------------------------------------
A C D E F G H I K L M N P Q R S T V W Y

1	0.008	0.001	0.001	0.007	0.002	0.004	0.01	0.006	0.14	0.006	0.005	0.009	0.002	0.01	0.7	0.009	0.01	0.006	0.002	0.003
2	...
...
----------------------------------------------------------------------

Empirical distribution mixture models. Developed by Dominik Schrempf (2019).

positional arguments:
  filenames             Names of files containing site distributions

optional arguments:
  -h, --help            show this help message and exit
  --transformation {none,clr,lclr}, -t {none,clr,lclr}
                        Transformation (default: None)
  -k K                  Number of clusters (default: 64)
  -p PREFIX, --prefix PREFIX
                        Prefix of output files. (default: None)

[1] Nguyen, L., Schmidt, H. A., von Haeseler, A., & Minh, B. Q. (2015).
Iq-tree: a fast and effective stochastic algorithm for estimating
maximum-likelihood phylogenies. Molecular Biology and Evolution, 32(1),
268–274. http://dx.doi.org/10.1093/molbev/msu300

[2] Lartillot, N., Rodrigue, N., Stubbs, D., & Richer, J. (2013). Phylobayes
mpi: phylogenetic reconstruction with infinite mixtures of profiles in a
parallel environment. Systematic Biology, 62(4), 611–615.
http://dx.doi.org/10.1093/sysbio/syt022

[3] Höhna, S., Landis, M. J., Heath, T. A., Boussau, B., Lartillot, N., Moore,
B. R., Huelsenbeck, J. P., … (2016). Revbayes: bayesian phylogenetic inference
using graphical models and an interactive model-specification language.
Systematic Biology, 65(4), 726–736. http://dx.doi.org/10.1093/sysbio/syw021

[4] Aitchison, J. (1982). The statistical analysis of compositional data.
Journal of the Royal Statistical Society, Series B (Methological), 44(2),
139–177.

[5] Godichon-Baggioni, A., Maugis-Rabusseau, C., & Rau, A. (2017).
Clustering transformed compositional data using K-means, with applications in
gene expression and bicycle sharing system data. ArXiv, 1–32.
#+end_example
